File: backend/models/clan_info.py
import re
from datetime import datetime
from models import db


class ClanInfo(db.Model):
    __tablename__ = 'clan_info'
    id = db.Column(db.Integer, primary_key=True)
    clan_id = db.Column(db.Integer, unique=True, nullable=False, index=True)
    name = db.Column(db.String(100), nullable=False)
    logo_url = db.Column(db.String(300), default='')
    description = db.Column(db.Text, default='')
    leader_nick = db.Column(db.String(100), default='')
    leader_rank = db.Column(db.String(100), default='')
    clan_rank = db.Column(db.String(100), default='')
    clan_level = db.Column(db.Integer, default=0)
    step = db.Column(db.Integer, default=0)
    talents = db.Column(db.Integer, default=0)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    def __repr__(self):
        return f'<ClanInfo {self.name}>'


class ClanMemberInfo(db.Model):
    __tablename__ = 'clan_member_info'
    id = db.Column(db.Integer, primary_key=True)
    clan_id = db.Column(db.Integer, db.ForeignKey('clan_info.clan_id'), nullable=False, index=True)
    nick = db.Column(db.String(100), nullable=False)
    icon = db.Column(db.String(10), default='')
    game_rank = db.Column(db.String(100), default='')
    level = db.Column(db.Integer, default=0)
    profession = db.Column(db.String(100), default='')
    profession_level = db.Column(db.Integer, default=0)
    clan_role = db.Column(db.String(100), default='')
    join_date = db.Column(db.String(20), default='')
    trial_until = db.Column(db.String(20), default='')
    is_deleted = db.Column(db.Boolean, default=False)

    clan = db.relationship('ClanInfo', foreign_keys=[clan_id], primaryjoin='ClanMemberInfo.clan_id == ClanInfo.clan_id')

    def __repr__(self):
        return f'<ClanMemberInfo {self.nick}>'


class ClanHierarchy(db.Model):
    __tablename__ = 'clan_hierarchy'
    id = db.Column(db.Integer, primary_key=True)
    clan_id = db.Column(db.Integer, db.ForeignKey('clan_info.clan_id'), nullable=False, index=True)
    role_name = db.Column(db.String(100), nullable=False)
    level = db.Column(db.Integer, default=0)
    color = db.Column(db.String(20), default='#00d4aa')
    icon = db.Column(db.String(10), default='')
    sort_order = db.Column(db.Integer, default=0)
    can_invite = db.Column(db.Boolean, default=False)
    can_kick = db.Column(db.Boolean, default=False)
    can_edit = db.Column(db.Boolean, default=False)
    can_analyze = db.Column(db.Boolean, default=True)
    min_level = db.Column(db.Integer, default=0)

    clan = db.relationship('ClanInfo', foreign_keys=[clan_id], primaryjoin='ClanHierarchy.clan_id == ClanInfo.clan_id')

    def __repr__(self):
        return f'<ClanHierarchy {self.role_name}>'

File: frontend/src/types/clanInfo.ts
export interface ClanInfoData {
  clan_id: number;
  name: string;
  logo_url: string;
  logo_big: string;
  logo_small: string;
  description: string;
  leader_nick: string;
  leader_rank: string;
  clan_rank: string;
  clan_level: number;
  step: number;
  talents: number;
  total_players: number;
  current_players: number;
  council: string[];
  clan_structure: ClanStructure;
  structure_warning?: string;
  updated_at: string;
}

export interface ClanStructure {
  leader?: {
    nick: string;
    description: string;
  };
  deputies?: Array<{
    nick: string;
    description: string;
  }>;
  council?: Array<{
    nick: string;
    description: string;
  }>;
  commander?: {
    nick: string;
    description: string;
  };
  has_members?: boolean;
  council_slots?: number;
}

export interface ClanMemberData {
  id?: number;
  nick: string;
  icon?: string;
  game_rank: string;
  level: number;
  profession: string;
  profession_level: number;
  clan_role: string;
  join_date: string;
  trial_until: string;
  is_deleted?: boolean;
  left_date?: string;
  leave_reason?: string;
}

export interface LeftMemberData {
  id: number;
  nick: string;
  icon?: string;
  game_rank: string;
  level: number;
  profession: string;
  profession_level: number;
  clan_role: string;
  join_date: string;
  left_date: string;
  leave_reason: string;
}

export interface TreasuryOperationData {
  id: number;
  date: string;
  nick: string;
  operation_type: string;
  object_name: string;
  quantity: number;
  compensation_flag: boolean;
  compensation_comment: string;
}
